zilliqa/tests/it/staking.rs

```diff
@@ -0,0 +1,233 @@
+use std::ops::DerefMut;
+
+use ethabi::Token;
+use ethers::{
+    middleware::SignerMiddleware,
+    providers::{Middleware, Provider},
+    signers::LocalWallet,
+    types::{BlockId, BlockNumber, TransactionRequest},
+};
+use primitive_types::H160;
+use tracing::{info, trace};
+use zilliqa::{contracts, crypto::NodePublicKey, state::contract_addr};
+
+use crate::{LocalRpcClient, Network};
+
+async fn check_miner_got_reward(
+    wallet: &SignerMiddleware<Provider<LocalRpcClient>, LocalWallet>,
+    block: impl Into<BlockId> + Send + Sync,
+) {
+    let block = wallet.get_block(block).await.unwrap().unwrap();
+    let miner = block.author.unwrap();
+    let balance_before = wallet
+        .get_balance(miner, Some((block.number.unwrap() - 1).into()))
+        .await
+        .unwrap();
+    let balance_after = wallet
+        .get_balance(miner, Some(block.number.unwrap().into()))
+        .await
+        .unwrap();
+
+    assert!(balance_before < balance_after);
+}
+
+async fn deposit_stake(
+    network: &mut Network,
+    wallet: &SignerMiddleware<Provider<LocalRpcClient>, LocalWallet>,
+    key: NodePublicKey,
+    stake: u128,
+    reward_address: H160,
+) {
+    // Transfer the new validator enough ZIL to stake.
+    let tx = TransactionRequest::pay(reward_address, stake);
+    let hash = wallet.send_transaction(tx, None).await.unwrap().tx_hash();
+    network
+        .run_until_async(
+            || async {
+                wallet
+                    .get_transaction_receipt(hash)
+                    .await
+                    .unwrap()
+                    .is_some()
+            },
+            50,
+        )
+        .await
+        .unwrap();
```

Comment:
I recently added a `network.run_until_receipt(Wallet, Hash)` helper which makes cases like this a lot more readable, since it's extremely boilerplate.
(And for reference also similarly also a `run_until_block(Wallet, U64)`.)

So it'd just be
```suggestion
    network.run_until_receipt(&wallet, hash, 50).await;
```

And similarly below

Reply:
Thanks, that's indeed very useful!